src/CommonDBTM.php

```diff
@@ -667,32 +667,31 @@ public function updateInDB($updates, $oldvalues = [])
     {
         global $DB;
 
+        $tobeupdated = [];
         foreach ($updates as $field) {
             if (isset($this->fields[$field])) {
-                $DB->update(
-                    $this->getTable(),
-                    [$field => $this->fields[$field]],
-                    ['id' => $this->fields['id']]
-                );
-                if ($DB->affectedRows() == 0) {
-                    if (isset($oldvalues[$field])) {
-                        unset($oldvalues[$field]);
-                    }
-                }
-            } else {
-               // Clean oldvalues
-                if (isset($oldvalues[$field])) {
+                if (isset($oldvalues[$field]) && $this->fields[$field] == $oldvalues[$field]) {
                     unset($oldvalues[$field]);
                 }
+                $tobeupdated[$field] = $this->fields[$field];
+            } else {
+                // Clean oldvalues
+                unset($oldvalues[$field]);
             }
         }
+        $DB->update(
+            $this->getTable(),
+            $tobeupdated,
+            ['id' => $this->fields['id']]
+        );
+        $affected_rows = $DB->affectedRows();
 
-        if (count($oldvalues)) {
+        if (count($oldvalues) && $affected_rows > 0) {
             Log::constructHistory($this, $oldvalues, $this->fields);
             $this->getFromDB($this->fields[$this->getIndexName()]);
         }
```

Comment:
Maybe it should be safer to reload the object even if there was not affected rows. Indeed, `$this->fields` may have been overriden in some hook code and may not reflect the real state in DB. Reloading the object will, in this case, ensure that `$this->fields` correspond to real values.
```suggestion
        }
        $this->getFromDB($this->fields[$this->getIndexName()]);
```

Reply:
I'm not sure an extra query if there were no changes is required. Also, this changes existing behavior.